packages/tenant-process/src/services/validators.ts

```diff
@@ -46,6 +50,98 @@ export function assertVerifiedAttributeExistsInTenant(
   }
 }
 
+export async function assertVerifiedAttributeOperationAllowed({
+  producerId,
+  consumerId,
+  attributeId,
+  agreementStates,
+  readModelService,
+  error,
+  limit,
+  offset,
+}: {
+  producerId: string;
+  consumerId: string;
+  attributeId: string;
+  agreementStates: AgreementState[];
+  readModelService: ReadModelService;
+  error: Error;
+  limit: number;
+  offset: number;
+}): Promise<void> {
+  // Get agreements
+  const agreements = await readModelService.getAgreements(
+    { producerId, consumerId, agreementStates },
+    limit,
+    offset
+  );
+
+  // Extract descriptor IDs
+  const descriptorIds = agreements.results.map(
+    (agreement) => agreement.descriptorId
+  );
+
+  // Get eServices concurrently
+  const eServices = (
+    await Promise.all(
+      agreements.results.map((agreement) =>
+        readModelService.getEServiceById(agreement.eserviceId)
+      )
+    )
+  ).filter((eService) => eService !== undefined) as EService[];
```

Comment:
```suggestion
  ).filter((eService): eService is EService => eService !== undefined);
```
You can use this as a type guard in order to be sure that eServices doesn't have any undefined values